File: frontend/src/api/network.ts
import client, { ApiResponse } from './client';

// Types
export interface NetworkInterface {
  name: string;
  index: number;
  hardwareAddr: string;
  flags: string[];
  mtu: number;
  addresses: string[];
  isUp: boolean;
  speed: string;
  type: string;
}

export interface InterfaceStats {
  name: string;
  rxBytes: number;
  txBytes: number;
  rxPackets: number;
  txPackets: number;
  rxErrors: number;
  txErrors: number;
  rxDropped: number;
  txDropped: number;
}

export interface Route {
  destination: string;
  gateway: string;
  genmask: string;
  flags: string;
  metric: number;
  iface: string;
}

export interface DNSConfig {
  nameservers: string[];
  searchDomains: string[];
}

export interface FirewallRule {
  number: number;
  action: string;
  from: string;
  to: string;
  protocol?: string;
  port?: string;
  description?: string;
}

export interface FirewallStatus {
  enabled: boolean;
  defaultIncoming: string;
  defaultOutgoing: string;
  defaultRouted: string;
  rules: FirewallRule[];
}

export interface NetworkBridge {
  id: string;
  name: string;
  description: string;
  ports: string;
  // IPv4
  ip_address?: string;
  gateway?: string;
  // IPv6 (Proxmox-style)
  ipv6_address?: string;
  ipv6_gateway?: string;
  // Bridge settings
  vlan_aware?: boolean;
  autostart: boolean;
  status: string;
  last_error?: string;
  // Pending changes
  has_pending_changes: boolean;
  pending_ports?: string;
  pending_ip_address?: string;
  pending_gateway?: string;
  pending_ipv6_address?: string;
  pending_ipv6_gateway?: string;
  pending_vlan_aware?: boolean;
  created_at: string;
  updated_at: string;
}

export interface PendingNetworkChange {
  id: string;
  change_type: string;
  action: string;
  resource_id: string;
  current_config?: string;
  pending_config: string;
  description?: string;
  created_by?: string;
  priority: number;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface PendingChangesResponse {
  has_pending: boolean;
  count: number;
  changes: PendingNetworkChange[];
}

export interface DiagnosticResult {
  command: string;
  output: string;
  success: boolean;
  error?: string;
}

// API
export const networkApi = {
  // Interfaces
  async listInterfaces(): Promise<ApiResponse<NetworkInterface[]>> {
    const response = await client.get('/network/interfaces');
    return response.data;
  },

  async getInterfaceStats(): Promise<ApiResponse<InterfaceStats[]>> {
    const response = await client.get('/network/interfaces/stats');
    return response.data;
  },

  async setInterfaceState(name: string, state: 'up' | 'down'): Promise<ApiResponse<any>> {
    const response = await client.post(`/network/interfaces/${name}/state`, { state });
    return response.data;
  },

  async configureInterface(
    name: string,
    mode: 'static' | 'dhcp',
    address?: string,
    netmask?: string,
    gateway?: string
  ): Promise<ApiResponse<any>> {
    const response = await client.post(`/network/interfaces/${name}/configure`, {
      mode,
      address,
      netmask,
      gateway,
    });
    return response.data;
  },

  // Routes
  async getRoutes(): Promise<ApiResponse<Route[]>> {
    const response = await client.get('/network/routes');
    return response.data;
  },

  async addRoute(destination: string, gateway?: string, iface?: string, metric?: number): Promise<ApiResponse<any>> {
    const response = await client.post('/network/routes', {
      destination,
      gateway,
      interface: iface,
      metric,
    });
    return response.data;
  },

  async deleteRoute(destination: string, gateway?: string, iface?: string): Promise<ApiResponse<any>> {
    const response = await client.delete('/network/routes', {
      data: {
        destination,
        gateway,
        interface: iface,
      },
    });
    return response.data;
  },

  // DNS
  async getDNS(): Promise<ApiResponse<DNSConfig>> {
    const response = await client.get('/network/dns');
    return response.data;
  },

  async setDNS(nameservers: string[], searchDomains: string[]): Promise<ApiResponse<any>> {
    const response = await client.post('/network/dns', { nameservers, searchDomains });
    return response.data;
  },

  // Firewall
  async getFirewallStatus(): Promise<ApiResponse<FirewallStatus>> {
    const response = await client.get('/network/firewall');
    return response.data;
  },

  async setFirewallState(enabled: boolean): Promise<ApiResponse<any>> {
    const response = await client.post('/network/firewall/state', { enabled });
    return response.data;
  },

  async addFirewallRule(
    action: string,
    port: string,
    protocol: string,
    from: string,
    to: string
  ): Promise<ApiResponse<any>> {
    const response = await client.post('/network/firewall/rules', { action, port, protocol, from, to });
    return response.data;
  },

  async deleteFirewallRule(number: number): Promise<ApiResponse<any>> {
    const response = await client.delete(`/network/firewall/rules/${number}`);
    return response.data;
  },

  async setDefaultPolicy(direction: string, policy: string): Promise<ApiResponse<any>> {
    const response = await client.post('/network/firewall/default', { direction, policy });
    return response.data;
  },

  async resetFirewall(): Promise<ApiResponse<any>> {
    const response = await client.post('/network/firewall/reset', {});
    return response.data;
  },

  // Diagnostics
  async ping(host: string, count: number = 4): Promise<ApiResponse<DiagnosticResult>> {
    const response = await client.post('/network/diagnostics/ping', { host, count });
    return response.data;
  },

  async traceroute(host: string): Promise<ApiResponse<DiagnosticResult>> {
    const response = await client.post('/network/diagnostics/traceroute', { host });
    return response.data;
  },

  async netstat(options: string = ''): Promise<ApiResponse<DiagnosticResult>> {
    const response = await client.post('/network/diagnostics/netstat', { options });
    return response.data;
  },

  // Wake-on-LAN
  async wakeOnLAN(macAddress: string): Promise<ApiResponse<any>> {
    const response = await client.post('/network/wol', { macAddress });
    return response.data;
  },

  // Bridge management
  async listBridges(): Promise<ApiResponse<string[]>> {
    const response = await client.get('/network/bridges');
    return response.data;
  },

  async createBridge(name: string, ports: string[]): Promise<ApiResponse<any>> {
    const response = await client.post('/network/bridges', { name, ports });
    return response.data;
  },

  async deleteBridge(name: string): Promise<ApiResponse<any>> {
    const response = await client.delete(`/network/bridges/${name}`);
    return response.data;
  },

  async attachPortToBridge(bridgeName: string, port: string): Promise<ApiResponse<any>> {
    const response = await client.post(`/network/bridges/${bridgeName}/attach`, { port });
    return response.data;
  },

  async detachPortFromBridge(port: string): Promise<ApiResponse<any>> {
    const response = await client.post(`/network/bridges/detach`, { port });
    return response.data;
  },

  // Proxmox-style pending changes workflow
  async getPendingChanges(): Promise<ApiResponse<PendingChangesResponse>> {
    const response = await client.get('/network/pending-changes');
    return response.data;
  },

  async createBridgeWithPendingChanges(
    name: string,
    description: string,
    ports: string[],
    ipAddress?: string,
    gateway?: string,
    ipv6Address?: string,
    ipv6Gateway?: string,
    vlanAware: boolean = false,
    autostart: boolean = true
  ): Promise<ApiResponse<NetworkBridge>> {
    const response = await client.post('/network/bridges/pending', {
      name,
      description,
      ports,
      ip_address: ipAddress,
      gateway,
      ipv6_address: ipv6Address,
      ipv6_gateway: ipv6Gateway,
      vlan_aware: vlanAware,
      autostart,
    });
    return response.data;
  },

  async updateBridgeWithPendingChanges(
    name: string,
    description?: string,
    ports?: string[],
    ipAddress?: string,
    gateway?: string,
    ipv6Address?: string,
    ipv6Gateway?: string,
    vlanAware?: boolean,
    autostart?: boolean
  ): Promise<ApiResponse<NetworkBridge>> {
    const response = await client.put(`/network/bridges/${name}/pending`, {
      description,
      ports,
      ip_address: ipAddress,
      gateway,
      ipv6_address: ipv6Address,
      ipv6_gateway: ipv6Gateway,
      vlan_aware: vlanAware,
      autostart,
    });
    return response.data;
  },

  async updateInterfaceWithPendingChanges(
    name: string,
    ipAddress?: string,
    gateway?: string,
    ipv6Address?: string,
    ipv6Gateway?: string,
    autostart: boolean = true,
    comment?: string
  ): Promise<ApiResponse<any>> {
    const response = await client.put(`/network/interfaces/${name}/pending`, {
      ip_address: ipAddress,
      gateway,
      ipv6_address: ipv6Address,
      ipv6_gateway: ipv6Gateway,
      autostart,
      comment,
    });
    return response.data;
  },

  async applyPendingChanges(): Promise<ApiResponse<any>> {
    const response = await client.post('/network/apply-changes', {});
    return response.data;
  },

  async discardPendingChanges(changeId?: string): Promise<ApiResponse<any>> {
    const response = await client.post('/network/discard-changes', { change_id: changeId });
    return response.data;
  },

  async rollbackToSnapshot(snapshotId: string): Promise<ApiResponse<any>> {
    const response = await client.post('/network/rollback', { snapshot_id: snapshotId });
    return response.data;
  },

  async getStoredBridges(): Promise<ApiResponse<NetworkBridge[]>> {
    const response = await client.get('/network/bridges/stored');
    return response.data;
  },
};

File: backend/internal/api/handlers/network.go
// Revision: 2025-11-16 | Author: Claude | Version: 1.1.1
package handlers

import (
	"encoding/json"
	"net/http"
	"strconv"

	"github.com/Stumpf-works/stumpfworks-nas/internal/network"
	"github.com/Stumpf-works/stumpfworks-nas/pkg/errors"
	"github.com/Stumpf-works/stumpfworks-nas/pkg/utils"

	"github.com/go-chi/chi/v5"
)

// NetworkHandler handles network-related requests
type NetworkHandler struct{}

// NewNetworkHandler creates a new network handler
func NewNetworkHandler() *NetworkHandler {
	return &NetworkHandler{}
}

// ListInterfaces handles GET /api/network/interfaces
func (h *NetworkHandler) ListInterfaces(w http.ResponseWriter, r *http.Request) {
	interfaces, err := network.ListInterfaces()
	if err != nil {
		utils.RespondError(w, errors.InternalServerError("Failed to list interfaces", err))
		return
	}

	utils.RespondSuccess(w, interfaces)
}

// GetInterfaceStats handles GET /api/network/interfaces/stats
func (h *NetworkHandler) GetInterfaceStats(w http.ResponseWriter, r *http.Request) {
	stats, err := network.GetInterfaceStats()
	if err != nil {
		utils.RespondError(w, errors.InternalServerError("Failed to get interface stats", err))
		return
	}

	utils.RespondSuccess(w, stats)
}

// SetInterfaceState handles POST /api/network/interfaces/{name}/state
func (h *NetworkHandler) SetInterfaceState(w http.ResponseWriter, r *http.Request) {
	name := chi.URLParam(r, "name")

	var req struct {
		State string `json:"state"` // "up" or "down"
	}

	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		utils.RespondError(w, errors.BadRequest("Invalid request", err))
		return
	}

	var err error
	if req.State == "up" {
		err = network.SetInterfaceUp(name)
	} else if req.State == "down" {
		err = network.SetInterfaceDown(name)
	} else {
		utils.RespondError(w, errors.BadRequest("Invalid state", nil))
		return
	}

	if err != nil {
		utils.RespondError(w, errors.InternalServerError("Failed to set interface state", err))
		return
	}

	utils.RespondSuccess(w, map[string]string{"message": "Interface state updated"})
}

// ConfigureInterface handles POST /api/network/interfaces/{name}/configure
func (h *NetworkHandler) ConfigureInterface(w http.ResponseWriter, r *http.Request) {
	name := chi.URLParam(r, "name")

	var req struct {
		Mode    string `json:"mode"`    // "static" or "dhcp"
		Address string `json:"address"` // for static
		Netmask string `json:"netmask"` // for static
		Gateway string `json:"gateway"` // for static
	}

	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		utils.RespondError(w, errors.BadRequest("Invalid request", err))
		return
	}

	var err error
	if req.Mode == "static" {
		if req.Address == "" || req.Netmask == "" {
			utils.RespondError(w, errors.BadRequest("Missing required fields", nil))
			return
		}
		err = network.ConfigureStaticIP(name, req.Address, req.Netmask, req.Gateway)
	} else if req.Mode == "dhcp" {
		err = network.ConfigureDHCP(name)
	} else {
		utils.RespondError(w, errors.BadRequest("Invalid mode", nil))
		return
	}

	if err != nil {
		utils.RespondError(w, errors.InternalServerError("Failed to configure interface", err))
		return
	}

	utils.RespondSuccess(w, map[string]string{"message": "Interface configured successfully"})
}

// GetRoutes handles GET /api/network/routes
func (h *NetworkHandler) GetRoutes(w http.ResponseWriter, r *http.Request) {
	routes, err := network.GetRoutes()
	if err != nil {
		utils.RespondError(w, errors.InternalServerError("Failed to get routes", err))
		return
	}

	utils.RespondSuccess(w, routes)
}

// AddRoute handles POST /api/network/routes
func (h *NetworkHandler) AddRoute(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Destination string `json:"destination"`
		Gateway     string `json:"gateway"`
		Interface   string `json:"interface"`
		Metric      int    `json:"metric"`
	}

	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		utils.RespondError(w, errors.BadRequest("Invalid request body", err))
		return
	}

	// Validate destination
	if req.Destination == "" {
		utils.RespondError(w, errors.BadRequest("Destination is required", nil))
		return
	}

	// Either gateway or interface must be provided
	if req.Gateway == "" && req.Interface == "" {
		utils.RespondError(w, errors.BadRequest("Either gateway or interface must be provided", nil))
		return
	}

	if err := network.AddRoute(req.Destination, req.Gateway, req.Interface, req.Metric); err != nil {
		utils.RespondError(w, errors.InternalServerError("Failed to add route", err))
		return
	}

	utils.RespondSuccess(w, map[string]string{"message": "Route added successfully"})
}

// DeleteRoute handles DELETE /api/network/routes
func (h *NetworkHandler) DeleteRoute(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Destination string `json:"destination"`
		Gateway     string `json:"gateway"`
		Interface   string `json:"interface"`
	}

	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		utils.RespondError(w, errors.BadRequest("Invalid request body", err))
		return
	}

	// Validate destination
	if req.Destination == "" {
		utils.RespondError(w, errors.BadRequest("Destination is required", nil))
		return
	}

	if err := network.DeleteRoute(req.Destination, req.Gateway, req.Interface); err != nil {
		utils.RespondError(w, errors.InternalServerError("Failed to delete route", err))
		return
	}

	utils.RespondSuccess(w, map[string]string{"message": "Route deleted successfully"})
}

// GetDNS handles GET /api/network/dns
func (h *NetworkHandler) GetDNS(w http.ResponseWriter, r *http.Request) {
	config, err := network.GetDNSConfig()
	if err != nil {
		utils.RespondError(w, errors.InternalServerError("Failed to get DNS config", err))
		return
	}

	utils.RespondSuccess(w, config)
}

// SetDNS handles POST /api/network/dns
func (h *NetworkHandler) SetDNS(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Nameservers   []string `json:"nameservers"`
		SearchDomains []string `json:"searchDomains"`
	}

	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		utils.RespondError(w, errors.BadRequest("Invalid request", err))
		return
	}

	if err := network.SetDNSConfig(req.Nameservers, req.SearchDomains); err != nil {
		utils.RespondError(w, errors.InternalServerError("Failed to set DNS config", err))
		return
	}

	utils.RespondSuccess(w, map[string]string{"message": "DNS configuration updated"})
}

// GetFirewallStatus handles GET /api/network/firewall
func (h *NetworkHandler) GetFirewallStatus(w http.ResponseWriter, r *http.Request) {
	status, err := network.GetFirewallStatus()
	if err != nil {
		utils.RespondError(w, errors.InternalServerError("Failed to get firewall status", err))
		return
	}

	utils.RespondSuccess(w, status)
}

// SetFirewallState handles POST /api/network/firewall/state
func (h *NetworkHandler) SetFirewallState(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Enabled bool `json:"enabled"`
	}

	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		utils.RespondError(w, errors.BadRequest("Invalid request", err))
		return
	}

	var err error
	if req.Enabled {
		err = network.EnableFirewall()
	} else {
		err = network.DisableFirewall()
	}

	if err != nil {
		utils.RespondError(w, errors.InternalServerError("Failed to set firewall state", err))
		return
	}

	utils.RespondSuccess(w, map[string]string{"message": "Firewall state updated"})
}

// AddFirewallRule handles POST /api/network/firewall/rules
func (h *NetworkHandler) AddFirewallRule(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Action   string `json:"action"`   // allow, deny, reject
		Port     string `json:"port"`
		Protocol string `json:"protocol"` // tcp, udp
		From     string `json:"from"`
		To       string `json:"to"`
	}

	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		utils.RespondError(w, errors.BadRequest("Invalid request", err))
		return
	}

	if err := network.AddFirewallRule(req.Action, req.Port, req.Protocol, req.From, req.To); err != nil {
		utils.RespondError(w, errors.InternalServerError("Failed to add firewall rule", err))
		return
	}

	utils.RespondSuccess(w, map[string]string{"message": "Firewall rule added"})
}

// DeleteFirewallRule handles DELETE /api/network/firewall/rules/{number}
func (h *NetworkHandler) DeleteFirewallRule(w http.ResponseWriter, r *http.Request) {
	numberStr := chi.URLParam(r, "number")
	number, err := strconv.Atoi(numberStr)
	if err != nil {
		utils.RespondError(w, errors.BadRequest("Invalid rule number", err))
		return
	}

	if err := network.DeleteFirewallRule(number); err != nil {
		utils.RespondError(w, errors.InternalServerError("Failed to delete firewall rule", err))
		return
	}

	utils.RespondSuccess(w, map[string]string{"message": "Firewall rule deleted"})
}

// SetDefaultPolicy handles POST /api/network/firewall/default
func (h *NetworkHandler) SetDefaultPolicy(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Direction string `json:"direction"` // incoming, outgoing, routed
		Policy    string `json:"policy"`    // allow, deny, reject
	}

	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		utils.RespondError(w, errors.BadRequest("Invalid request", err))
		return
	}

	if err := network.SetDefaultPolicy(req.Direction, req.Policy); err != nil {
		utils.RespondError(w, errors.InternalServerError("Failed to set default policy", err))
		return
	}

	utils.RespondSuccess(w, map[string]string{"message": "Default policy updated"})
}

// ResetFirewall handles POST /api/network/firewall/reset
func (h *NetworkHandler) ResetFirewall(w http.ResponseWriter, r *http.Request) {
	if err := network.ResetFirewall(); err != nil {
		utils.RespondError(w, errors.InternalServerError("Failed to reset firewall", err))
		return
	}

	utils.RespondSuccess(w, map[string]string{"message": "Firewall reset successfully"})
}

// Ping handles POST /api/network/diagnostics/ping
func (h *NetworkHandler) Ping(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Host  string `json:"host"`
		Count int    `json:"count"`
	}

	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		utils.RespondError(w, errors.BadRequest("Invalid request", err))
		return
	}

	if req.Count <= 0 {
		req.Count = 4
	}

	result, err := network.Ping(req.Host, req.Count)
	if err != nil {
		utils.RespondError(w, errors.InternalServerError("Failed to ping", err))
		return
	}

	utils.RespondSuccess(w, result)
}

// Traceroute handles POST /api/network/diagnostics/traceroute
func (h *NetworkHandler) Traceroute(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Host string `json:"host"`
	}

	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		utils.RespondError(w, errors.BadRequest("Invalid request", err))
		return
	}

	result, err := network.Traceroute(req.Host)
	if err != nil {
		utils.RespondError(w, errors.InternalServerError("Failed to traceroute", err))
		return
	}

	utils.RespondSuccess(w, result)
}

// Netstat handles POST /api/network/diagnostics/netstat
func (h *NetworkHandler) Netstat(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Options string `json:"options"`
	}

	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		utils.RespondError(w, errors.BadRequest("Invalid request", err))
		return
	}

	result, err := network.Netstat(req.Options)
	if err != nil {
		utils.RespondError(w, errors.InternalServerError("Failed to run netstat", err))
		return
	}

	utils.RespondSuccess(w, result)
}

// WakeOnLAN handles POST /api/network/wol
func (h *NetworkHandler) WakeOnLAN(w http.ResponseWriter, r *http.Request) {
	var req struct {
		MacAddress string `json:"macAddress"`
	}

	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		utils.RespondError(w, errors.BadRequest("Invalid request", err))
		return
	}

	if err := network.WakeOnLAN(req.MacAddress); err != nil {
		utils.RespondError(w, errors.InternalServerError("Failed to send WOL packet", err))
		return
	}

	utils.RespondSuccess(w, map[string]string{"message": "Wake-on-LAN packet sent"})
}
